go/nbs/table_set.go

```diff
@@ -142,15 +154,19 @@ func (ts tableSet) Prepend(mt *memTable, stats *Stats) tableSet {
 // compact the N smallest (by number of chunks) tables which can be compacted
 // into a new table such that upon replacing the N input tables, the
 // resulting table will still have the fewest chunks in the tableSet.
-func (ts tableSet) Compact(stats *Stats) (ns tableSet, compactees chunkSources) {
+func (ts tableSet) Compact(stats *Stats) tableSet {
 	t1 := time.Now()
 
-	ns = tableSet{
-		novel: make(chunkSources, len(ts.novel)),
-		p:     ts.p,
-		rl:    ts.rl,
+	ns := tableSet{
+		novel:      make(chunkSources, len(ts.novel)),
+		compacted:  make(chunkSources, len(ts.compacted)+1),
+		compactees: make(chunkSources, len(ts.compactees)),
+		p:          ts.p,
+		rl:         ts.rl,
 	}
 	copy(ns.novel, ts.novel)
+	copy(ns.compacted[1:], ts.compacted) // leave the first slot for the newly compacted table
```

Comment:
My primary thing was that, just looking at the `tableSet` interface, only being able to call `Compact` once felt like a weird and artificial restriction.
I started coding it that way anyway and discovered that having `compacted` be a `chunkSource` instead of `chunkSources` made the code, in my opinion, a little clunkier to read. Like, you have to do explicit `nil` checks instead of just ranging over a thing that might happen to be nil. Or, instead of being able to re-use a little helper function that works on `novel` and `upstream`, you have to do something slightly different. See, for example `has()` and `hasMany()`.

That's a minor consideration, certainly, but given that restricting it to just one `Compact()` call already felt kind of weird, I decided to do it this way. There seemed to be little downside and it made the code and API better.